lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+		bav.Snapshot, profilePKID, lockupDurationNanoSecs)
+
+	// Cache the DB entry in the in-memory map.
+	if lockupYieldCurvePoint != nil {
+		bav._setLockupYieldCurvePoint(lockupYieldCurvePoint)
+	}
+
+	return lockupYieldCurvePoint
+}
+
+func (bav *UtxoView) GetLocalYieldCurvePoints(profilePKID *PKID, lockupDuration int64) (
+	_leftLockupPoint *LockupYieldCurvePoint, _rightLockupPoint *LockupYieldCurvePoint) {
+	var leftLockupPoint *LockupYieldCurvePoint
+	var rightLockupPoint *LockupYieldCurvePoint
+
+	// Check the view for yield curve points.
+	if _, pointsInView := bav.PKIDToLockupYieldCurvePointKeyToLockupYieldCurvePoints[*profilePKID]; pointsInView {
+		for _, lockupYieldCurvePoint := range bav.PKIDToLockupYieldCurvePointKeyToLockupYieldCurvePoints[*profilePKID] {
+			// Check for nil pointer cases.
+			if lockupYieldCurvePoint.LockupDurationNanoSecs < lockupDuration && leftLockupPoint == nil {
+				leftLockupPoint = lockupYieldCurvePoint
+			}
+			if lockupYieldCurvePoint.LockupDurationNanoSecs >= lockupDuration && rightLockupPoint == nil {
+				rightLockupPoint = lockupYieldCurvePoint
+			}
+
+			// Check if the point is "more left" than the current left point.
+			if lockupYieldCurvePoint.LockupDurationNanoSecs < lockupDuration &&
+				lockupYieldCurvePoint.LockupDurationNanoSecs > leftLockupPoint.LockupDurationNanoSecs {
+				leftLockupPoint = lockupYieldCurvePoint.Copy()
+			}
+
+			// Check if the point is "more right" than the current right point.
+			if lockupYieldCurvePoint.LockupDurationNanoSecs >= lockupDuration &&
+				lockupYieldCurvePoint.LockupDurationNanoSecs < leftLockupPoint.LockupDurationNanoSecs {
+				rightLockupPoint = lockupYieldCurvePoint.Copy()
+			}
+		}
+	}
+
+	// Now we quickly fetch left and right local yield curve points from the DB using careful seek operations.
+	leftDBLockupPoint, rightDBLockupPoint := DBGetLocalYieldCurvePoints(
+		bav.GetDbAdapter().badgerDb, bav.Snapshot, profilePKID, lockupDuration)
+
+	// Check for nil pointer cases.
+	if leftDBLockupPoint != nil &&
+		leftDBLockupPoint.LockupDurationNanoSecs < lockupDuration {
+		leftLockupPoint = leftDBLockupPoint
+	}
+	if rightDBLockupPoint != nil &&
+		rightDBLockupPoint.LockupDurationNanoSecs >= lockupDuration {
+		rightLockupPoint = rightDBLockupPoint
+	}
+
+	// Check for an updated left and right yield curve point from the DB.
+	if leftDBLockupPoint != nil &&
+		leftDBLockupPoint.ProfilePKID.Eq(profilePKID) &&
+		leftDBLockupPoint.LockupDurationNanoSecs < lockupDuration &&
+		leftDBLockupPoint.LockupDurationNanoSecs > leftLockupPoint.LockupDurationNanoSecs {
+		leftLockupPoint = leftDBLockupPoint
+	}
+	if rightDBLockupPoint != nil &&
+		rightDBLockupPoint.ProfilePKID.Eq(profilePKID) &&
+		rightDBLockupPoint.LockupDurationNanoSecs >= lockupDuration &&
+		rightDBLockupPoint.LockupDurationNanoSecs < rightLockupPoint.LockupDurationNanoSecs {
+		rightLockupPoint = rightDBLockupPoint
+	}
+
+	return leftLockupPoint, rightLockupPoint
```

Comment:
There is a weird edge-case here where something could be in the view marked as isDeleted. If this is the case, then you need to actually do multiple things to avoid getting screwed by it:
1. You need to skip it when iterating over the view, which I didn't see above
2. You need to skip it when you fetch from the db. This is actually harder than it seems because you don't have a way of getting the *next* leftmost/rightmost point while *skipping* a point that's in the view.

Given all of the complexity associated with this left/right stuff, I had an alternative proposal for you. Let's just do the inefficient thing of fetching *all* the points from the db every time, and get rid of this left/right stuff in Badger. The new code would look as follows:
* When you call _get, you read *all* the lockup points from the db for a particular PKID and set them on the view if they're not already there.
* When you need leftmost/rightmost, you call the above _get to load up the view with *everything* in badger, and then iterate over the view like you have above (but just be careful to check for isDeleted)

The above is much simpler, and avoids having to write skipping logic in badger. If it ever becomes too inefficient, we can always replace it with something better.

Reply:
Btw in case it isn't clear, the above change would keep the exact schema you have for the view where it's `map[PKID]map[LockupPointKey]` or whatever.